decode/lwfile.1
." $Id: lwfile.1,v 1.8 1996/05/10 17:32:49 yoksas Exp $
.TH LWFILE 1 "$Date: 1996/05/10 17:32:49 $" "Printed: \n(yr.\n(mo.\n(dy" "UNIDATA UTILITIES"
.SH NAME
lwfile \- Unidata LDM decoder for McIDAS "textual" data
.SH SYNOPSIS
.HP
.ft B
lwfile
.nh
\%[-v]
\%[-x]
\%[-l\ \fIlogpath\fP]
\%[-[f|F]\ \fIfile\fP]
.hy
.ft
.SH DESCRIPTION
.LP
This program is an LDM decoder for McIDAS text data (data \fBnot\fP of type 
\fBAREA\fP, \fBGRID\fP, or \fBMD\fP; the FSL, netCDF, wind profiler 
file, PROFILER.CDF, is an example):
it reads from standard input or from a file and decodes that data
in the manner of the McIDAS \fBLWFILE\fP action (which is simply to save the
data, with no interpretation, into a file).
.LP
This program can be invoked by an entry in an LDM
configuration file similar to the following:
.sp
.RS
.ta \w'MCIDAS   'u
.nf
MCIDAS	^(LWFILE .*)
	PIPE
	-close /usr/local/ldm/bin/lwfile
.fi
.RE
.LP
Additional, profusely-annotated examples of LDM5 pqact.conf entries can be
found in the file ldm-mcidas-pqact.conf in the \fBetc/\fP subdirectory of the 
ldm-mcidas installation tree.
.fi
.RE
.LP
Since this program reads from standard input, it can also be invoked as a
stand-alone program, outside the LDM environment.
.SH OPTIONS
.TP
.BI "-d " datadir
Change to directory \fIdatadir\fP before decoding.  As a consequence, the
output file will appear in that directory and the \fBROUTE.SYS\fP and
\fBSYSKEY.TAB\fP files will be looked for and, if found, used, and updated
in that directory.  Note that \fBROUTE.SYS\fP and \fBSYSKEY.TAB\fP must be
both readable and writable by the user running \fBlwfile\fP.
.TP
.BI "-f " path
Read from file \fIpath\fP rather than from standard input.
.TP
.BI "-F " path
Read from file \fIpath\fP rather than from standard input.  At exit, delete
the file.
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and diagnostic
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.
.TP
.B -v
Verbose logging.
Informative messages (level LOG_INFO) are logged.
By default, only messages of severity LOG_NOTICE and greater are
logged. See the Unidata  LDM Site  Manager's  Guide  for more information on 
message logging.
.TP
.B -x
Debug logging.
Debugging messages (level LOG_DEBUG) are logged. See the Unidata  LDM Site  
Manager's  Guide  for more information on message logging.
.SH NOTES
.LP
Unidata recommends that the following McIDAS-X routing table update commands
be use to setup \fBROUTE.SYS\fP entries used by \fBlwfile\fP (for more
information see the section on \fBROUTE\fP in the McIDAS-X Users Guide):
 
.nf
ROUTE ADD U4 TEXT X X PROFILER.CDF CC=3 SYS=2041 2042 2043 PP=PROFILER.BAT "FSL Wind profiler
ROUTE ADD UM TEXT X X ADMIN.MSG CC=1 "Administrative Message
ROUTE ADD US TEXT X X UNIDATAS CC=1 "Undecoded SAO Data
ROUTE REL U4 UM US
.fi

\fBROUTE.SYS\fP will be used by \fBlwfile\fP if it exists in the directory
in which \fBlwfile\fP creates its output \fBTEXT\fP files (see \fB-d\fP flag above)
and is both readable and writable by the user running \fBlwfile\fP.

\fBSYSKEY.TAB\fP entries specified by the \fBSYS=\fP keywords listed above
will be updated if \fBSYSKEY.TAB\fP exists in the directory in which
\fBlwfile\fP creates its output \fBTEXT\fP files (see \fB-d\fP flag above) and
is both readable and writable by the user running \fBlwfile\fP.
.RE
.SH "SEE ALSO"
.LP
.BR ~ldm-mcidas/etc/ldm-mcidas-pqact.conf,
.BR ldmd (1),
.BR mcingest (1),
.BR mctingest (1),
.BR ulog (3),
.BR syslogd (8).


RELEASE_NOTES
VERSION	COMMENTS
-------	-------------------------------------------------------------

2012    2012-01-18

        Updated version number to 2012 to reflect small updates (e.g.,
        new src/port/cfortran/cfortran.h, modified src/decode/Makefile)
        and that the package is still viable.

2008    2009-04-30

        Note: The version number is kept at 2008 since McIDAS went
        into maintenance mode with v2008.

        A variety of changes to C routines, Makefiles and test data
        files.

        Introduction of a new McIDAS lightning decoder, 'lgt2md', that
        handles both NLDN binary and USPLN ASCII data record formats.
        'lgt2md' is designed as a drop-in replacement for 'nldn2md':
        user's pattern action file entries for decoding NLDN data
        can be easily updated by replacing 'nldn2md' with 'lgt2md'.

        ROUTE.SYS (distributed as route.sys) was updated to include
        a new product, 'LP', for USPLN lightning data.

        The Makefile in the src/decode directory was reorganized to
        build the subset of the encoders/decoders contained in the package
        that are written entirely in C:

        area2png    - PNG-compresses the image portion of single-banded
                      McIDAS AREA files
        pnga2area   - uncompresses images created by 'area2png'
        zlibg2gini  - uncompresses the Zlib-compressed images contained
                      in the NOAAPort broadcast (and conveyed in the
                      IDD in the NIMAGE feed)
        pngg2gini   - uncompress NOAAPort images that have been
                      PNG-compressed by the Unidata NOAAPort ingest
                      package (archaic)
        proftomd    - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files
        lgt2md      - decodes NLDN and USPLN lightning data into
                      McIDAS MD files

        This set is built by running 'make' or 'make basic'.
        
        In addition to this set, the Fortran-based 'nldn2md' NLDN
        lightning decoder will be built by 'make full'.

        The Makefile in src/decode also provides for the (attempted)
        building of archaic McIDAS decoders:

        cdftomd     - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files (use 'proftomd' instead)
        gunrv2      - decodes compressed McIDAS GRID files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      GRID format
        lwfile
        lwtmd2      - decodes compressed McIDAS MD files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      MD format
        lwtoa3      - decodes delta-compressed McIDAS image files from
                      the old Zephyr/Alden satellite broadcast into McIDAS
                      AREA format
        nids2area   - converts Nexrad Level III images into McIDAS
                      AREA file format (best to use McIDAS IMGCOPY from
                      NEXR to AREA datasets directly as McIDAS capabilities
                      are kept up to date)
 
        In addition decoders built by 'make full', an attepmt will be
        made to build these decoders when the user runs 'make all'.
        There is _no_ guarantee that these decoders will be built
        successfully!

2004    2004-09-14

        Added O_TRUNC flag to mkdirs_open call in pnga2area.c,
        pngg2gini.c, and zlibg2gini.c.  This corrects a problem
        encountered when decoding Unidata-Wisconsin images (IDD
        feedtype UNIWISC aka MCIDAS) when using McIDAS routing table
        (ROUTE.SYS) concepts.  The problem was manifested in decoded
        image files keeping the size of the images they were
        overwriting.  So, the file size was always the "high water
        mark" of sizes for the products.

        NOTES:
        - this did not affect the sizes of files decoded without using
          McIDAS routing table concepts
        - the decoded images using McIDAS routing table concepts were
          valid/usable even though they are larger than they need to
          be

        Changed IDD datastream name from MCIDAS to UNIWISC in example
        pqact.conf files.

        Updated distribution with the versions of SATANNOT, SATBAND,
        and SCHEMA from the latest Unidata McIDAS distribution, v2004.

        Cleaned up batch.k shell script that runs McIDAS postprocess
        BATCH files.  Internal comments should now be easier to follow
        than in previous versions.

        Updated man pages for currently used decoders: area2png(1),
        nldn2md(1), pnga2area(1), pngg2gini(1), proftomd(1), and
        zlibg2gini(1).

        Removed building of decoders now considered to be obsolete:
        cdftomd(1), lwtoa3(1), gunrv2(1), lwfile(1), lwtmd2(1),
        and nids2area(1).  Source code for these decoders is still
        included in the distribution, but their use is not guaranteed
        mainly since subroutines they use from the McIDAS library
        are no longer supported.

2003    2003-07-29

        Added utililty routine zlibg2gini that uncompresses Zlib
        compressed GINI images in NIMAGE IDD datastream.

        Fixed a bug that kept McIDAS routing table, ROUTE.SYS,
        dates and times from being updated.

2002    2002-12-20

        Replaced information messages in pngg2gini.c that referred to
        pnga2area with pngg2gini.

        Corrected logic in pnga2area.c for processing a product whose
        product code is not contained in the McIDAS routing table,
        ROUTE.SYS.  Before, the routing table would get corrupted
        and the AREA file number included in the '-r pcode,anum'
        flag would not be used.  Now, the routing table will not
        be updated for unregistered products, and the default AREA
        file number will be used.
	

decode/lwfile.c
/*
 * $Id: lwfile.c,v 1.5 1996/04/29 18:18:01 steve Exp $
 *
 * Perform the McIDAS `LWFILE' action on standard input.
 */

#include <udposix.h>

#include <stddef.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>

#include "cfortran.h"
#include "ulog.h"
#include "mcidas.h"

#ifndef PATH_MAX
#   define PATH_MAX	_POSIX_PATH_MAX
#endif

#include "mc_proto.h"

/*
 * McIDAS FORTRAN routines used by this module:
 */
PROTOCCALLSFFUN3(INT,GETRTE,getrte,INT,PINT,PSTRING)
#define GETRTE(in,out,str) \
    CCALLSFFUN3(GETRTE,getrte,INT,PINT,PSTRING,in,out,str)
#define FSHRTE(irte,istat) \
    CCALLSFSUB2(FSHRTE,fshrte,INT,INT,irte,istat)
PROTOCCALLSFFUN1(INT,NKWP,nkwp,STRING)
#define NKWP(str) \
    CCALLSFFUN1(NKWP,nkwp,STRING,str)


/*
 * Perform the McIDAS action.
 */
    void
main0_()
{
    const char	*outpath;		/* output file pathname */

    /*
     * Establish output file pathname.
     */
    if (Mccmdstr(NULL, 1, "LWFILE.DAT", &outpath) < 0)
	uerror(
	    "Mccmdstr() failure: Couldn't get pathname of LWFILE output file");
    else
    {
	int		    irte	= 0;
	int		    in_no	= 0;
	int		    out_no;

	if (NKWP("DIALPROD") != 0)
	    irte	= GETRTE(in_no, out_no, outpath);

	if (irte >= 0) {
	    int		status	= -2;	/* return status (total failure) */
	    FILE	*outfp;

	    /*
	     * Open the output file.
	     */
	    if ((outfp = fopen(outpath, "w")) == NULL) {
		serror("Can't open output file \"%s\"", outpath);

	    } else {
		static struct mcpacket pkt;

		status	= -1;
		if (mc_datacopy(&pkt, outfp, INT_MAX) >= 0)
		    status	= 1;

		if (fclose(outfp) == EOF) {
		    serror("Can't close output file \"%s\"", outpath);
		    status	= -1;
		}
	    }				/* output file opened */

	    if (irte > 0)
		FSHRTE(irte, status);
	}					/* GETRTE => decode product */
    }						/* outpath obtained */
}
